docs/custom-transfers.md

```diff
@@ -284,6 +349,320 @@ to the stdin of the transfer process:
+```
+
+Fields:
+* `event`: Always "bulk-complete" to identify this message
+* `oid`: The bulk transfer ID
+* `error`: Should contain a `code` and `message` explaining the error and `retry` flag indicating whether this bulk should be retried
+
+`retry` flag in case of bulks corresponds to retrying the whole bulk regardless of any previous 
+successfull results for individual artifacts. A bulk can be reported as sucessfull though
+some artifacts can be retried later via the retry flag in the bulk item completion messages. 
+
+The bulk transfer completion error never preceeds the item completion messages, and it
+is sent only once after all items have been processed.
+
+The bulk transfer failure is not considered a fatal error, and the process can continue 
+to handle other bulks or items.
+
+### Stage 3: Finish & Cleanup
+
+When all transfers have been processed, git-lfs will send the following message
+to the stdin of the transfer process:
+
+```json
+{ "event": "terminate" }
+```
+
+On receiving this message the transfer process should clean up and terminate.
+No response is expected.
+
+## Protocol Flow Example
+
+Here's an example flow for downloading 2 files in a bulk:
+
+### Client to Process:
+```json
+{"event": "init", "operation": "download", "remote": "origin", "concurrent": true, "concurrenttransfers": 3}
+{"event": "bulk-header", "oid": "bulk-001", "size": 2 }
+{"event": "download", "oid": "sha256:file1", "size": 1024, "path": "", "action": {"href": "https://example.com/file1"}}
+{"event": "download", "oid": "sha256:file2", "size": 2048, "path": "", "action": {"href": "https://example.com/file2"}}
+{"event": "bulk-footer", "oid": "bulk-001", "size": 3072, "path": "", "action": null}
+```
+
+### Process to Client (Sequential Processing):
+```json
+{ }
+{"event": "progress", "oid": "bulk-001", "bytesSoFar": 0, "bytesSinceLast": 0}
+{"event": "progress", "oid": "bulk-001", "bytesSoFar": 1024, "bytesSinceLast": 1024}
+{"event": "complete", "oid": "sha256:file1", "path": "/tmp/file1" }
+{"event": "progress", "oid": "bulk-001", "bytesSoFar": 3072, "bytesSinceLast": 2048}
+{"event": "complete", "oid": "sha256:file2", "path": "/tmp/file2" }
+{"event": "bulk-complete", "oid": "bulk-001" }
+```
+
+### Process to Client (Parallel Processing):
+```json
+{ }
+{"event": "progress", "oid": "bulk-001", "bytesSoFar": 0, "bytesSinceLast": 0}
+{"event": "progress", "oid": "bulk-001", "bytesSoFar": 2048, "bytesSinceLast": 2048}
+{"event": "complete", "oid": "sha256:file2", "path": "/tmp/file2" }
+{"event": "progress", "oid": "bulk-001", "bytesSoFar": 3072, "bytesSinceLast": 1024}
+{"event": "complete", "oid": "sha256:file1", "path": "/tmp/file1" }
+{"event": "bulk-complete", "oid": "bulk-001" }
+```
+
+Note: In the parallel example, file2 completes before file1, demonstrating that 
+completion order is independent of the order in which items were sent.
```

Comment:
Rather than including a "parallel processing" example here, I think it would be useful to show or explain that additional bulk transfers can occur before the final `terminate` message.

The client could just start with another `bulk-header`/`bulk-footer` message pair with a series of individual object `upload` or `download` messages between them, and then process the responses, repeating as necessary before finally closing the connection with a `terminate` message.

As for the concept of parallel processing, I think we should simply explain that any parallelism is controlled by the `concurrenttransfers` settings exactly as for the `basic` mode.